File: samples/generators/jpa/src/main/javagen/topmodel/jpa/sample/demo/entities/utilisateur/Utilisateur.java
////
//// ATTENTION CE FICHIER EST GENERE AUTOMATIQUEMENT !
////

package topmodel.jpa.sample.demo.entities.utilisateur;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import jakarta.annotation.Generated;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EntityListeners;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.OneToOne;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;

import topmodel.jpa.sample.demo.entities.securite.Profil;

/**
 * Utilisateur de l'application.
 */
@Generated("TopModel : https://github.com/klee-contrib/topmodel")
@Entity
@Table(name = "UTILISATEUR", uniqueConstraints = {
    @UniqueConstraint(columnNames = {"UTI_EMAIL","UTI_ID_PARENT"})})
@EntityListeners(AuditingEntityListener.class)
public class Utilisateur {

	/**
	 * Id technique.
	 */
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	@Id
	@Column(name = "UTI_ID", nullable = false, columnDefinition = "int")
	private Integer id;

	/**
	 * Age en années de l'utilisateur.
	 */
	@Column(name = "UTI_AGE", nullable = true, precision = 20, scale = 9, columnDefinition = "numeric")
	private BigDecimal age = 6;

	/**
	 * Profil de l'utilisateur.
	 */
	@ManyToOne(fetch = FetchType.LAZY, optional = true, targetEntity = Profil.class)
	@JoinColumn(name = "PRO_ID", referencedColumnName = "PRO_ID")
	private Profil profil;

	/**
	 * Email de l'utilisateur.
	 */
	@Column(name = "UTI_EMAIL", nullable = true, length = 50, columnDefinition = "varchar")
	private String email;

	/**
	 * Nom de l'utilisateur.
	 */
	@Column(name = "UTI_NOM", nullable = true, length = 3, columnDefinition = "varchar")
	private String nom = "Jabx";

	/**
	 * Si l'utilisateur est actif.
	 */
	@Column(name = "UTI_ACTIF", nullable = true, columnDefinition = "boolean")
	private Boolean actif;

	/**
	 * Type d'utilisateur en Many to one.
	 */
	@ManyToOne(fetch = FetchType.LAZY, optional = true, targetEntity = TypeUtilisateur.class)
	@JoinColumn(name = "TUT_CODE", referencedColumnName = "TUT_CODE")
	private TypeUtilisateur typeUtilisateur = new TypeUtilisateur(TypeUtilisateurCode.ADM);

	/**
	 * Utilisateur parent.
	 */
	@OneToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL, optional = true)
	@JoinColumn(name = "UTI_ID_PARENT", referencedColumnName = "UTI_ID", unique = true)
	private Utilisateur utilisateurParent;

	/**
	 * Utilisateur enfants.
	 */
	@OneToMany(cascade = CascadeType.ALL, fetch = FetchType.LAZY, mappedBy = "utilisateurEnfant")
	private List<Utilisateur> utilisateursEnfant;

	/**
	 * Date de création de l'utilisateur.
	 */
	@Column(name = "UTI_DATE_CREATION", nullable = true, columnDefinition = "date")
	@CreatedDate
	private LocalDate dateCreation;

	/**
	 * Date de modification de l'utilisateur.
	 */
	@Column(name = "UTI_DATE_MODIFICATION", nullable = true, columnDefinition = "date")
	@LastModifiedDate
	private LocalDateTime dateModification;

	/**
	 * Association réciproque de Utilisateur.UtilisateursEnfant.
	 */
	@ManyToOne(fetch = FetchType.LAZY, optional = true, targetEntity = Utilisateur.class)
	@JoinColumn(name = "UTI_ID_ENFANT", referencedColumnName = "UTI_ID")
	private Utilisateur utilisateurEnfant;

	/**
	 * No arg constructor.
	 */
	public Utilisateur() {
	}

	/**
	 * Getter for id.
	 *
	 * @return value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#id id}.
	 */
	public Integer getId() {
		return this.id;
	}

	/**
	 * Getter for age.
	 *
	 * @return value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#age age}.
	 */
	public BigDecimal getAge() {
		return this.age;
	}

	/**
	 * Getter for profil.
	 *
	 * @return value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#profil profil}.
	 */
	public Profil getProfil() {
		return this.profil;
	}

	/**
	 * Getter for email.
	 *
	 * @return value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#email email}.
	 */
	public String getEmail() {
		return this.email;
	}

	/**
	 * Getter for nom.
	 *
	 * @return value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#nom nom}.
	 */
	public String getNom() {
		return this.nom;
	}

	/**
	 * Getter for actif.
	 *
	 * @return value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#actif actif}.
	 */
	public Boolean getActif() {
		return this.actif;
	}

	/**
	 * Getter for typeUtilisateur.
	 *
	 * @return value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#typeUtilisateur typeUtilisateur}.
	 */
	public TypeUtilisateur getTypeUtilisateur() {
		return this.typeUtilisateur;
	}

	/**
	 * Getter for utilisateurParent.
	 *
	 * @return value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#utilisateurParent utilisateurParent}.
	 */
	public Utilisateur getUtilisateurParent() {
		return this.utilisateurParent;
	}

	/**
	 * Getter for utilisateursEnfant.
	 *
	 * @return value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#utilisateursEnfant utilisateursEnfant}.
	 */
	public List<Utilisateur> getUtilisateursEnfant() {
		if(this.utilisateursEnfant == null)
			this.utilisateursEnfant = new ArrayList<>();
		return this.utilisateursEnfant;
	}

	/**
	 * Getter for dateCreation.
	 *
	 * @return value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#dateCreation dateCreation}.
	 */
	public LocalDate getDateCreation() {
		return this.dateCreation;
	}

	/**
	 * Getter for dateModification.
	 *
	 * @return value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#dateModification dateModification}.
	 */
	public LocalDateTime getDateModification() {
		return this.dateModification;
	}

	/**
	 * Getter for utilisateurEnfant.
	 *
	 * @return value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#utilisateurEnfant utilisateurEnfant}.
	 */
	public Utilisateur getUtilisateurEnfant() {
		return this.utilisateurEnfant;
	}

	/**
	 * Set the value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#id id}.
	 * @param id value to set
	 */
	public void setId(Integer id) {
		this.id = id;
	}

	/**
	 * Set the value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#age age}.
	 * @param age value to set
	 */
	public void setAge(BigDecimal age) {
		this.age = age;
	}

	/**
	 * Set the value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#profil profil}.
	 * @param profil value to set
	 */
	public void setProfil(Profil profil) {
		this.profil = profil;
	}

	/**
	 * Set the value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#email email}.
	 * @param email value to set
	 */
	public void setEmail(String email) {
		this.email = email;
	}

	/**
	 * Set the value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#nom nom}.
	 * @param nom value to set
	 */
	public void setNom(String nom) {
		this.nom = nom;
	}

	/**
	 * Set the value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#actif actif}.
	 * @param actif value to set
	 */
	public void setActif(Boolean actif) {
		this.actif = actif;
	}

	/**
	 * Set the value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#typeUtilisateur typeUtilisateur}.
	 * @param typeUtilisateur value to set
	 */
	public void setTypeUtilisateur(TypeUtilisateur typeUtilisateur) {
		this.typeUtilisateur = typeUtilisateur;
	}

	/**
	 * Set the value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#utilisateurParent utilisateurParent}.
	 * @param utilisateurParent value to set
	 */
	public void setUtilisateurParent(Utilisateur utilisateurParent) {
		this.utilisateurParent = utilisateurParent;
	}

	/**
	 * Set the value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#utilisateursEnfant utilisateursEnfant}.
	 * @param utilisateursEnfant value to set
	 */
	public void setUtilisateursEnfant(List<Utilisateur> utilisateursEnfant) {
		this.utilisateursEnfant = utilisateursEnfant;
	}

	/**
	 * Set the value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#dateCreation dateCreation}.
	 * @param dateCreation value to set
	 */
	public void setDateCreation(LocalDate dateCreation) {
		this.dateCreation = dateCreation;
	}

	/**
	 * Set the value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#dateModification dateModification}.
	 * @param dateModification value to set
	 */
	public void setDateModification(LocalDateTime dateModification) {
		this.dateModification = dateModification;
	}

	/**
	 * Set the value of {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur#utilisateurEnfant utilisateurEnfant}.
	 * @param utilisateurEnfant value to set
	 */
	public void setUtilisateurEnfant(Utilisateur utilisateurEnfant) {
		this.utilisateurEnfant = utilisateurEnfant;
	}

	/**
	 * Enumération des champs de la classe {@link topmodel.jpa.sample.demo.entities.utilisateur.Utilisateur Utilisateur}.
	 */
	public enum Fields  {
        ID(Integer.class), //
        AGE(BigDecimal.class), //
        PROFIL(Profil.class), //
        EMAIL(String.class), //
        NOM(String.class), //
        ACTIF(Boolean.class), //
        TYPE_UTILISATEUR(TypeUtilisateur.class), //
        UTILISATEUR_PARENT(Utilisateur.class), //
        UTILISATEURS_ENFANT(List.class), //
        DATE_CREATION(LocalDate.class), //
        DATE_MODIFICATION(LocalDateTime.class), //
        UTILISATEUR_ENFANT(Utilisateur.class);

		private Class<?> type;

		private Fields(Class<?> type) {
			this.type = type;
		}

		public Class<?> getType() {
			return this.type;
		}
	}
}

File: samples/generators/php/src/Entity/Utilisateur/Utilisateur.php
<?php
////
//// ATTENTION CE FICHIER EST GENERE AUTOMATIQUEMENT !
////

namespace App\Entity\Utilisateur;

use App\Entity\Securite\Profil;
use App\Repository\Utilisateur\UtilisateurRepository;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\ORM\Mapping\Column;
use Doctrine\ORM\Mapping\Entity;
use Doctrine\ORM\Mapping\GeneratedValue;
use Doctrine\ORM\Mapping\Id;
use Doctrine\ORM\Mapping\JoinColumn;
use Doctrine\ORM\Mapping\ManyToOne;
use Doctrine\ORM\Mapping\OneToMany;
use Doctrine\ORM\Mapping\OneToOne;
use Doctrine\ORM\Mapping\SequenceGenerator;
use Doctrine\ORM\Mapping\Table;
use Doctrine\ORM\Mapping\UniqueConstraint;

#[Entity(repositoryClass: UtilisateurRepository::class)]
#[Table(name: 'UTILISATEUR')]
#[UniqueConstraint(name: "UTI_EMAIL_UTI_ID_PARENT_UNIQ", columns: ["UTI_EMAIL, UTI_ID_PARENT"])]
class Utilisateur
{
  #[Id]
  #[GeneratedValue(strategy: "SEQUENCE")]
  #[SequenceGenerator(sequenceName: "SEQ_UTILISATEUR")]
  #[Column(name: 'UTI_ID')]
  private int $id;

  #[Column(name: 'UTI_AGE', length: 20, scale: 9, nullable: true)]
  private int|null $age = 6l;

  #[ManyToOne(targetEntity: Profil::class)]
  #[JoinColumn(name: 'PRO_ID', referencedColumnName: 'PRO_ID')]
  private Profil $profil;

  #[Column(name: 'UTI_EMAIL', length: 50, nullable: true)]
  private string|null $email;

  #[Column(name: 'UTI_NOM', length: 3, nullable: true)]
  private string|null $nom = "Jabx";

  #[Column(name: 'UTI_ACTIF', nullable: true)]
  private bool|null $actif;

  #[ManyToOne(targetEntity: TypeUtilisateur::class)]
  #[JoinColumn(name: 'TUT_CODE', referencedColumnName: 'TUT_CODE')]
  private TypeUtilisateur $typeUtilisateur;

  #[OneToOne(targetEntity: Utilisateur::class)]
  #[JoinColumn(name: 'UTI_ID_PARENT', referencedColumnName: 'UTI_ID')]
  private Utilisateur $utilisateurParent;

  /**
   * @var Collection<Utilisateur>
   */
  #[OneToMany(mappedBy: 'utilisateurEnfant', targetEntity: Utilisateur::class)]
  private Collection $utilisateursEnfant;

  #[Column(name: 'UTI_DATE_CREATION', nullable: true)]
  private Date|null $dateCreation;

  #[Column(name: 'UTI_DATE_MODIFICATION', nullable: true)]
  private Date|null $dateModification;

  #[ManyToOne(targetEntity: Utilisateur::class)]
  #[JoinColumn(name: 'UTI_ID_ENFANT', referencedColumnName: 'UTI_ID')]
  private Utilisateur $utilisateurEnfant;

  public function __construct()
  {
    $this->utilisateursEnfant = new ArrayCollection();
  }

  public function getId(): int
  {
    return $this->id;
  }

  public function getAge(): int|null
  {
    return $this->age;
  }

  public function getProfil(): Profil|null
  {
    return $this->profil;
  }

  public function getEmail(): string|null
  {
    return $this->email;
  }

  public function getNom(): string|null
  {
    return $this->nom;
  }

  public function getActif(): bool|null
  {
    return $this->actif;
  }

  public function getTypeUtilisateur(): TypeUtilisateur|null
  {
    return $this->typeUtilisateur;
  }

  public function getUtilisateurParent(): Utilisateur|null
  {
    return $this->utilisateurParent;
  }

  /**
   * @return Collection<Utilisateur>|null
   */
  public function getUtilisateursEnfant(): Collection|null
  {
    return $this->utilisateursEnfant;
  }

  public function getDateCreation(): Date|null
  {
    return $this->dateCreation;
  }

  public function getDateModification(): Date|null
  {
    return $this->dateModification;
  }

  public function getUtilisateurEnfant(): Utilisateur|null
  {
    return $this->utilisateurEnfant;
  }

  public function setId(int|null $id): self
  {
    $this->id = $id;

    return $this;
  }

  public function setAge(int|null $age): self
  {
    $this->age = $age;

    return $this;
  }

  public function setProfil(Profil|null $profil): self
  {
    $this->profil = $profil;

    return $this;
  }

  public function setEmail(string|null $email): self
  {
    $this->email = $email;

    return $this;
  }

  public function setNom(string|null $nom): self
  {
    $this->nom = $nom;

    return $this;
  }

  public function setActif(bool|null $actif): self
  {
    $this->actif = $actif;

    return $this;
  }

  public function setTypeUtilisateur(TypeUtilisateur|null $typeUtilisateur): self
  {
    $this->typeUtilisateur = $typeUtilisateur;

    return $this;
  }

  public function setUtilisateurParent(Utilisateur|null $utilisateurParent): self
  {
    $this->utilisateurParent = $utilisateurParent;

    return $this;
  }

  /**
   * @param Collection<Utilisateur>|null $utilisateursEnfant
   */
  public function setUtilisateursEnfant(Collection|null $utilisateursEnfant): self
  {
    $this->utilisateursEnfant = $utilisateursEnfant;

    return $this;
  }

  public function setDateCreation(Date|null $dateCreation): self
  {
    $this->dateCreation = $dateCreation;

    return $this;
  }

  public function setDateModification(Date|null $dateModification): self
  {
    $this->dateModification = $dateModification;

    return $this;
  }

  public function setUtilisateurEnfant(Utilisateur|null $utilisateurEnfant): self
  {
    $this->utilisateurEnfant = $utilisateurEnfant;

    return $this;
  }
}
